libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+        const { runner, status, consumers, changedProjectsOnly } = runnerContext;
+        const buildCacheContext: IOperationBuildCacheContext | undefined =
+          this._getBuildCacheContextByRunner(runner);
+
+        let blockCacheWrite: boolean = !buildCacheContext?.isCacheWriteAllowed;
+        let blockSkip: boolean = !buildCacheContext?.isSkipAllowed;
+
+        switch (status) {
+          case OperationStatus.Skipped: {
+            // Skipping means cannot guarantee integrity, so prevent cache writes in dependents.
+            blockCacheWrite = true;
+            break;
+          }
+
+          case OperationStatus.SuccessWithWarning:
+          case OperationStatus.Success: {
+            // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+            blockSkip ||= !changedProjectsOnly;
+            break;
+          }
+        }
+
+        // Apply status changes to direct dependents
+        for (const item of consumers) {
+          const itemRunnerBuildCacheContext: IOperationBuildCacheContext | undefined =
+            this._getBuildCacheContextByRunner(item.runner);
+          if (itemRunnerBuildCacheContext) {
+            if (blockCacheWrite) {
+              itemRunnerBuildCacheContext.isCacheWriteAllowed = false;
+            }
+            if (blockSkip) {
+              itemRunnerBuildCacheContext.isSkipAllowed = false;
+            }
+          }
+        }
+      }
+    );
+
+    hooks.afterExecuteOperations.tapPromise(PLUGIN_NAME, async () => {
+      for (const { buildCacheProjectLogWritable } of this._buildCacheContextByOperationRunner.values()) {
+        buildCacheProjectLogWritable?.close();
+      }
+      this._buildCacheContextByOperationRunner.clear();
+    });
+  }
+
+  private _applyShellOperationRunner(runner: ShellOperationRunner, context: ICreateOperationsContext): void {
+    const {
+      buildCacheConfiguration,
+      cobuildConfiguration,
+      phaseSelection: selectedPhases,
+      projectChangeAnalyzer
+    } = context;
+    const { hooks } = runner;
+
+    const buildCacheContext: IOperationBuildCacheContext = this._getBuildCacheContextByRunnerOrThrow(runner);
+
+    hooks.beforeExecute.tapPromise(
+      PLUGIN_NAME,
+      async (beforeExecuteContext: IOperationRunnerBeforeExecuteContext) => {
+        const earlyReturnStatus: OperationStatus | undefined = await (async () => {
+          const {
+            context,
+            runner,
+            terminal,
+            lastProjectDeps,
+            projectDeps,
+            trackedProjectFiles,
+            logPath,
+            errorLogPath,
+            rushProject,
+            phase,
+            commandName,
+            commandToRun,
+            earlyReturnStatus
+          } = beforeExecuteContext;
+          if (earlyReturnStatus) {
+            // If there is existing early return status, we don't need to do anything
+            return earlyReturnStatus;
+          }
```

Comment:
This is probably standard behavior for users of this hook, in which case the hook should probably be an `AsyncSeriesBailHook` (returning any value other than `undefined` aborts the sequence)